### PR TITLE
K8S Bootstrap token generation. 

### PR DESCRIPTION
**Pull request by JOUNAIDSoufiane**
Upon hardware creation, a token is created and then used by the K8S worker to create the corresponding secret in k8s. Furthermore, the Balena worker sets this token as the now device-specific K3S_Token environment variable

### CODE REVIEW

**Comment by JOUNAIDSoufiane** [comment]
## Code to call the Kubernetes API:
`            core_v1 = kubernetes_client()

            secret_data = {
                "description": f"Bootstrap token generated by doni k8s worker",
                "token-id": token_id,
                "token-secret": token_secret,
                "expiration": "never",
                "usage-bootstrap-signing": "true",
                "usage-bootstrap-authentication": "true",
            }

            # Encode the secret data
            encoded_secret_data = {k: base64.b64encode(v.encode()).decode() for k, v in secret_data.items()}

            # Create the Kubernetes Secret object
            secret = client.V1Secret(
                api_version="v1",
                kind="Secret",
                metadata=client.V1ObjectMeta(name=f"bootstrap-token-{token_id}"),
                type="bootstrap.kubernetes.io/token",
                data=encoded_secret_data
            )

            created_secret = core_v1.create_namespaced_secret(namespace="kube-system", body=secret)

            LOG.debug(f"Created secret: {created_secret.to_str()}")`


## Response from Kubernetes API 

`2023-12-13 11:52:15.953 7 DEBUG kubernetes.client.rest [-] response body: {"kind":"Secret","apiVersion":"v1","metadata":{"name":"bootstrap-token-ee2ar8","namespace":"kube-system","uid":"44d9e0c6-e959-4613-96d2-41ecb7b3e65e","resourceVersion":"35602632","creationTimestamp":"2023-12-13T17:52:15Z","managedFields":[{"manager":"OpenAPI-Generator","operation":"Update","apiVersion":"v1","time":"2023-12-13T17:52:15Z","fieldsType":"FieldsV1","fieldsV1":{"f:data":{".":{},"f:description":{},"f:expiration":{},"f:token-id":{},"f:token-secret":{},"f:usage-bootstrap-authentication":{},"f:usage-bootstrap-signing":{}},"f:type":{}}}]},"data":{"description":"Qm9vdHN0cmFwIHRva2VuIGdlbmVyYXRlZCBieSBkb25pIGs4cyB3b3JrZXI=","expiration":"bmV2ZXI=","token-id":"ZWUyYXI4","token-secret":"bTZobHNsYjA4eWNjaXp0eg==","usage-bootstrap-authentication":"dHJ1ZQ==","usage-bootstrap-signing":"dHJ1ZQ=="},"type":"bootstrap.kubernetes.io/token"}`

## Kubectl output

`soufianej@dev01:/opt/chi-in-a-box$ sudo kubectl get secrets -n kube-system
E1213 11:58:27.683112  345321 memcache.go:287] couldn't get resource list for metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E1213 11:58:27.684375  345321 memcache.go:121] couldn't get resource list for metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E1213 11:58:27.686046  345321 memcache.go:121] couldn't get resource list for metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E1213 11:58:27.688046  345321 memcache.go:121] couldn't get resource list for metrics.k8s.io/v1beta1: the server is currently unable to handle the request
NAME                                                 TYPE                                  DATA   AGE
daemon-set-controller-token-tqp5c                    kubernetes.io/service-account-token   3      268d
disruption-controller-token-wkcsx                    kubernetes.io/service-account-token   3      268d
endpointslice-controller-token-56d82                 kubernetes.io/service-account-token   3      268d
root-ca-cert-publisher-token-fjvxm                   kubernetes.io/service-account-token   3      268d
endpoint-controller-token-4gszj                      kubernetes.io/service-account-token   3      268d
deployment-controller-token-md86c                    kubernetes.io/service-account-token   3      268d
ephemeral-volume-controller-token-8lhzm              kubernetes.io/service-account-token   3      268d
replication-controller-token-xwpl5                   kubernetes.io/service-account-token   3      268d
job-controller-token-mjbrb                           kubernetes.io/service-account-token   3      268d
persistent-volume-binder-token-798m7                 kubernetes.io/service-account-token   3      268d
pod-garbage-collector-token-9lkqx                    kubernetes.io/service-account-token   3      268d
horizontal-pod-autoscaler-token-gbnw4                kubernetes.io/service-account-token   3      268d
pv-protection-controller-token-qmmgd                 kubernetes.io/service-account-token   3      268d
endpointslicemirroring-controller-token-hfv4q        kubernetes.io/service-account-token   3      268d
dev01.edge.chameleoncloud.org.node-password.k3s      Opaque                                1      268d
namespace-controller-token-tt2sr                     kubernetes.io/service-account-token   3      268d
certificate-controller-token-d4vph                   kubernetes.io/service-account-token   3      268d
ttl-controller-token-sfg5q                           kubernetes.io/service-account-token   3      268d
node-controller-token-msxxb                          kubernetes.io/service-account-token   3      268d
coredns-token-hwsmk                                  kubernetes.io/service-account-token   3      268d
expand-controller-token-8kxxz                        kubernetes.io/service-account-token   3      268d
local-path-provisioner-service-account-token-2g75d   kubernetes.io/service-account-token   3      268d
ttl-after-finished-controller-token-wxmh6            kubernetes.io/service-account-token   3      268d
service-account-controller-token-s7b2w               kubernetes.io/service-account-token   3      268d
generic-garbage-collector-token-42pv9                kubernetes.io/service-account-token   3      268d
metrics-server-token-2xb9b                           kubernetes.io/service-account-token   3      268d
cronjob-controller-token-28dlf                       kubernetes.io/service-account-token   3      268d
clusterrole-aggregation-controller-token-znxmq       kubernetes.io/service-account-token   3      268d
replicaset-controller-token-6p7n8                    kubernetes.io/service-account-token   3      268d
statefulset-controller-token-86b24                   kubernetes.io/service-account-token   3      268d
attachdetach-controller-token-7wnhj                  kubernetes.io/service-account-token   3      268d
pvc-protection-controller-token-r5z8l                kubernetes.io/service-account-token   3      268d
resourcequota-controller-token-9pvjb                 kubernetes.io/service-account-token   3      268d
default-token-g56lt                                  kubernetes.io/service-account-token   3      268d
c274f4b07335.node-password.k3s                       Opaque                                1      166d
73d72f74f9af.node-password.k3s                       Opaque                                1      166d
f28fcbbde56f.node-password.k3s                       Opaque                                1      166d
787c3bf5c2e0.node-password.k3s                       Opaque                                1      166d
07cffa036fa0.node-password.k3s                       Opaque                                1      166d
7a8091270702.node-password.k3s                       Opaque                                1      166d
3d8d3e25eecc.node-password.k3s                       Opaque                                1      166d
6bb3aac9b749.node-password.k3s                       Opaque                                1      166d
c8b8415d186a.node-password.k3s                       Opaque                                1      166d
96c4c89ea191.node-password.k3s                       Opaque                                1      166d
k3s-serving                                          kubernetes.io/tls                     2      268d
977ee37075d9.node-password.k3s                       Opaque                                1      166d
971c67f3c932.node-password.k3s                       Opaque                                1      166d
a40de278a1ba.node-password.k3s                       Opaque                                1      166d
af30a4533395.node-password.k3s                       Opaque                                1      166d
492cb92.node-password.k3s                            Opaque                                1      166d
00aa69a.node-password.k3s                            Opaque                                1      166d
iot-jetson01.node-password.k3s                       Opaque                                1      166d
iot-rpi4-01.node-password.k3s                        Opaque                                1      166d
iot-rpi4-0023.node-password.k3s                      Opaque                                1      63d
iot-jetson11.node-password.k3s                       Opaque                                1      62d
iot-rpi4-0025.node-password.k3s                      Opaque                                1      61d
msh-dev-rpi4-01.node-password.k3s                    Opaque                                1      7d20h`


## Expected output:

`soufianej@dev01:~$ sudo kubectl get secrets -n kube-system
E1213 12:00:05.537668  350730 memcache.go:287] couldn't get resource list for metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E1213 12:00:05.539356  350730 memcache.go:121] couldn't get resource list for metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E1213 12:00:05.541513  350730 memcache.go:121] couldn't get resource list for metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E1213 12:00:05.542959  350730 memcache.go:121] couldn't get resource list for metrics.k8s.io/v1beta1: the server is currently unable to handle the request
NAME                                                 TYPE                                  DATA   AGE
daemon-set-controller-token-tqp5c                    kubernetes.io/service-account-token   3      268d
disruption-controller-token-wkcsx                    kubernetes.io/service-account-token   3      268d
endpointslice-controller-token-56d82                 kubernetes.io/service-account-token   3      268d
root-ca-cert-publisher-token-fjvxm                   kubernetes.io/service-account-token   3      268d
endpoint-controller-token-4gszj                      kubernetes.io/service-account-token   3      268d
deployment-controller-token-md86c                    kubernetes.io/service-account-token   3      268d
ephemeral-volume-controller-token-8lhzm              kubernetes.io/service-account-token   3      268d
replication-controller-token-xwpl5                   kubernetes.io/service-account-token   3      268d
job-controller-token-mjbrb                           kubernetes.io/service-account-token   3      268d
persistent-volume-binder-token-798m7                 kubernetes.io/service-account-token   3      268d
pod-garbage-collector-token-9lkqx                    kubernetes.io/service-account-token   3      268d
horizontal-pod-autoscaler-token-gbnw4                kubernetes.io/service-account-token   3      268d
pv-protection-controller-token-qmmgd                 kubernetes.io/service-account-token   3      268d
endpointslicemirroring-controller-token-hfv4q        kubernetes.io/service-account-token   3      268d
dev01.edge.chameleoncloud.org.node-password.k3s      Opaque                                1      268d
namespace-controller-token-tt2sr                     kubernetes.io/service-account-token   3      268d
certificate-controller-token-d4vph                   kubernetes.io/service-account-token   3      268d
ttl-controller-token-sfg5q                           kubernetes.io/service-account-token   3      268d
node-controller-token-msxxb                          kubernetes.io/service-account-token   3      268d
coredns-token-hwsmk                                  kubernetes.io/service-account-token   3      268d
expand-controller-token-8kxxz                        kubernetes.io/service-account-token   3      268d
local-path-provisioner-service-account-token-2g75d   kubernetes.io/service-account-token   3      268d
ttl-after-finished-controller-token-wxmh6            kubernetes.io/service-account-token   3      268d
service-account-controller-token-s7b2w               kubernetes.io/service-account-token   3      268d
generic-garbage-collector-token-42pv9                kubernetes.io/service-account-token   3      268d
metrics-server-token-2xb9b                           kubernetes.io/service-account-token   3      268d
cronjob-controller-token-28dlf                       kubernetes.io/service-account-token   3      268d
clusterrole-aggregation-controller-token-znxmq       kubernetes.io/service-account-token   3      268d
replicaset-controller-token-6p7n8                    kubernetes.io/service-account-token   3      268d
statefulset-controller-token-86b24                   kubernetes.io/service-account-token   3      268d
attachdetach-controller-token-7wnhj                  kubernetes.io/service-account-token   3      268d
pvc-protection-controller-token-r5z8l                kubernetes.io/service-account-token   3      268d
resourcequota-controller-token-9pvjb                 kubernetes.io/service-account-token   3      268d
default-token-g56lt                                  kubernetes.io/service-account-token   3      268d
c274f4b07335.node-password.k3s                       Opaque                                1      166d
73d72f74f9af.node-password.k3s                       Opaque                                1      166d
f28fcbbde56f.node-password.k3s                       Opaque                                1      166d
787c3bf5c2e0.node-password.k3s                       Opaque                                1      166d
07cffa036fa0.node-password.k3s                       Opaque                                1      166d
7a8091270702.node-password.k3s                       Opaque                                1      166d
3d8d3e25eecc.node-password.k3s                       Opaque                                1      166d
6bb3aac9b749.node-password.k3s                       Opaque                                1      166d
c8b8415d186a.node-password.k3s                       Opaque                                1      166d
96c4c89ea191.node-password.k3s                       Opaque                                1      166d
k3s-serving                                          kubernetes.io/tls                     2      268d
977ee37075d9.node-password.k3s                       Opaque                                1      166d
971c67f3c932.node-password.k3s                       Opaque                                1      166d
a40de278a1ba.node-password.k3s                       Opaque                                1      166d
af30a4533395.node-password.k3s                       Opaque                                1      166d
492cb92.node-password.k3s                            Opaque                                1      166d
00aa69a.node-password.k3s                            Opaque                                1      166d
iot-jetson01.node-password.k3s                       Opaque                                1      166d
iot-rpi4-01.node-password.k3s                        Opaque                                1      166d
iot-rpi4-0023.node-password.k3s                      Opaque                                1      63d
iot-jetson11.node-password.k3s                       Opaque                                1      62d
iot-rpi4-0025.node-password.k3s                      Opaque                                1      61d
msh-dev-rpi4-01.node-password.k3s                    Opaque                                1      7d20h
bootstrap-token-ee2ar8                               bootstrap.kubernetes.io/token         6      4s`

File: doni/objects/hardware.py [hunk 1]
```diff
@@ -60,7 +62,8 @@ def save(self, context: "RequestContext" = None):
             HardwareDuplicateName: if a hardware with the same name exists.
             HardwareNotFound: if the hardware does not exist.
         """
-        updates = self.obj_get_changes()
+        updates = self.obj_get_changes() # BUG: Does not detect changes to hardware properties
+        updates["properties"] = self.properties
```
Comment (msherman64) [inline doni/objects/hardware.py:66]: Can you extract this to its own commit with more info?

**Review by msherman64 — APPROVED** [review]
Overall LGTM, can you squash/rebase a bit so that we have:

* a commit for the new feature
* a commit(s) for related bugs that weren't part of the feature (like the update thing)

File: doni/objects/hardware.py (hunk 1 above)
Comment (JOUNAIDSoufiane) [reply doni/objects/hardware.py:66]: Done!
